### PR TITLE
Add automatic directory listing generation to HttpFileService

### PR DESCRIPTION
**Pull request by trustin**
Motivation:

It is sometimes useful if a user can browse the files hosted by an
`HttpFileService` during a development phase.

Modifications:

- Added `canList()` and `list()` to `HttpVfs`.
- Added `HttpFileServiceBuilder.autoIndex(boolean)` and
  `HttpFileServiceConfig.autoIndex()`.
- Implemented simplistic automatic directory listing generation.
- Added a new example that demonstrates the use of `HttpFileService` and
  `HttpFile`.
- Miscellaneous:
  - Added `slf4j-simple` to example dependencies.
  - Added the respective web site links for examples in `README.md`.
  - Rename the section 'Annotated HTTP Services' to 'Annotated
    services'.

Result:

- One more example.
- Easier to demonstrate and play with `HttpFileService`.

### CODE REVIEW

File: examples/README.md [hunk 1]
```diff
@@ -14,6 +16,11 @@
 
 - `spring-boot-webflux`
   - Learn how to make Armeria serve your Spring Boot reactive web application.
+  - See [Using Armeria with Spring WebFlux](https://line.github.io/armeria/advanced-spring-webflux-integration.html).
+
+- `static-files`
+  - Learn how to server static files.
```
Comment (minwoox) [inline examples/README.md]: serve?

File: core/src/main/java/com/linecorp/armeria/server/file/HttpFileService.java [hunk 2]
```diff
@@ -191,9 +195,31 @@ private HttpFile findFile(ServiceRequestContext ctx, HttpRequest req) throws IOE
...
+                final List<String> listing = config.vfs().list(decodedMappedPath);
+                final HttpData autoIndex =
+                        AutoIndex.listingToHtml(ctx.decodedPath(), decodedMappedPath, listing);
+                return HttpFileBuilder.of(autoIndex)
+                                      .addHeader(HttpHeaderNames.CONTENT_TYPE, MediaType.HTML_UTF_8)
+                                      .build();
+            }
+        } else {
+            // Redirect to the slash appended path if 1) /index.html exists or 2) it has a directory listing.
+            if (findFile(ctx, decodedMappedPath + "index.html", supportedEncodings) != null ||
```
[the remark this answers is not in the document]
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/server/file/HttpFileService.java]: Yes. Fixed. :sweat_smile:

File: examples/README.md (hunk 1 above)
Comment (trustin) [reply examples/README.md]: Thanks. Fixed.

File: core/src/main/java/com/linecorp/armeria/server/file/AutoIndex.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+            "<!DOCTYPE html>\n<html><head><meta charset=\"UTF-8\"><title>Directory listing: ";
+    private static final String PART2 =
+            "</title><style>" +
+            "body { font-family: sans-serif; }" +
+            ".container { display: flex; }" +
+            ".directory-listing { margin: 0 auto; min-width: 60%; }" +
+            ".directory-listing ul { padding-left: 1.75em; }" +
+            "</style></head><body>\n" +
+            "<div class=\"container\"><div class=\"directory-listing\"><h1>Directory listing: ";
+    private static final String PART3 = "</h1>\n<p class=\"\">";
```
[the remark this answers is not in the document]
Comment (trustin) [reply core/src/main/java/com/linecorp/armeria/server/file/AutoIndex.java]: Ah yeah. :sweat_smile:

File: site/src/sphinx/server-http-file.rst [hunk 4]
```diff
@@ -21,6 +25,26 @@ Use :api:`HttpFileService` to serve static files under a certain directory. :api
...
 
+Auto-generating directory listings
+----------------------------------
+
+You can configure :api:`HttpFileService` to generate the directory listing of the directories without
+an ``index.html`` file using the ``autoIndex(boolean)`` method in :api:`HttpFileServiceBuilder`.
+
+.. code-block:: java
+
+    HttpFileServiceBuilder fsb =
```
[the remark this answers is not in the document]
Comment (trustin) [reply site/src/sphinx/server-http-file.rst]: Added.

**Comment by trustin** [comment]
Thanks for reviewing, @minwoox!

**Pull request merged**